design/secretless-e2e-tests-dnd.md

```diff
@@ -0,0 +1,256 @@
+
+[ Test Runner ] is software that provides a conventional way for defining and running test cases. They typically come with capabilities to measure test metrics, test outcomes and code coverage. This will also allow historical tracking. An example of a test-runner is Cucumber. Test cases can be defined in Gherkin
+
+[ Service Engine ] uses [ Service Config ] to create an instance of a [ Service ] 
+
+[ Test Case ] specifies a test case and contains some annotations detailing required [ Service ]s 
+
+[ Test Runner ] gathers annotations from [ Service ]s from [ Test Case ]s and issues requests for services to the [ Service Engine ] before running all [ Test Case ].
+
+[ Service Engine ] uses [ Service Config ] to create an instance of a [ Service ].
+
+Here is an example of the components in this architecture working together.
+0. A service definition exists for conjur and mysql
+```bash
+# Conjur service definition.  This is supplied to the start_services function and defines the service to be started
+declare -A conjur_service_config
+conjur_service_config=(
+["service_name"]="conjur_abc",
+["config"]="conjur.yaml",
+["start"]="conjur_start.sh",
+["stop"]="conjur_stop.sh"
+)
+
+declare -A mysql_service_config
+mysql_service_config=(
+["service_name"]="mysql_abc",
+["config"]="mysql.yaml",
+["start"]="mysql_start.sh",
+["stop"]="mysql_stop.sh"
+)
+``` 
+1. There is a test case written in Gherkin for validating using the MySQL service connector with Conjur.
+```
+Feature: Guess the word
+  
+  @conjur @mysql @mysql-conjur
+  Scenario: Connection to MySQL
+    Given Secretless is running as a sidecar in Kubernetes
+    Then the app is able to connect to MYSQL on "localhost:332"
+
+```
+2. Cucumber will register that there's a test case that requires Conjur, MySQL and MySQL credentials in Conjur.
+3. The developer calls `cucumber` to start the testss
+4. As a result of the annotations Cucumber will first call the engine to create the services.
+5. The configuration values
+ 
+TODO: provide examples of terms above
+TODO: split out components of the architecture by owner (infra, project developer etc.)
+
+## Interfaces
+
+TBD
+
+### Opaque API for lifecycle management of dependency resources
+
+The API makes it possible to implement service lifecycle methods in whatever way makes sense. That could be literal bash, docker-compose, calling an HTTP API to grab the service from a pool etc. This makes it possible to have similar experience between different environments (Jenkins, developer machine etc.). The same exact commands would be called on Jenkins as on the developer machine. The action of the commands would depend on the environment. For example, the developer machine can use a less resource intensive implementation of the service and Jenkins can run something that more closely approximates customer environments.
+
+Steps towards implementation
```

Comment:
just commenting here that `kubernetes-conjur-deploy` has uses outside of our test cases (we literally [advise people](https://docs.cyberark.com/Product-Doc/OnlineHelp/AAM-DAP/Latest/en/Content/Integrations/ConjurDeployFollowers.htm?Highlight=conjur-deploy#DeployfollowerstoKubernetes) to use this repo to deploy followers in K8s)

in addition, `kubernetes-conjur-demo` has a use case as a demo environment, we've just polluted it by also using it for e2e tests.

I have a strong preference to creating a separate repo instead of `demo` for secretless e2e tests. behind the scenes it may leverage `kubernetes-conjur-deploy` at first, but eventually the tools behind the service interface can leverage whatever makes sense (and should also probably live someplaces besides the `deploy` repo so that we can someday deprecate it)